### PR TITLE
Adding schema type checking support for OPA (opa eval -s)

### PR DESCRIPTION
**Pull request by aavarghese**
New support to input a JSON schema file via "opa eval --schema" that helps improve static type checking and returns precise error reports as you develop Rego code. Also, uses modified version of https://github.com/xeipuuv/gojsonschema from internal/ to help set Rego types, using gojsonschema's Compiler results. More details below:

Documentation for new "schema" feature in https://github.com/aavarghese/opa/blob/typeCheck/docs/content/schemas.md

Examples in https://github.com/aavarghese/opa-schema-examples

Co-authored-by: @vazirim Mandana Vaziri <mvaziri@us.ibm.com>
Co-authored-by: @aavarghese Ansu Varghese <avarghese@us.ibm.com>

### CODE REVIEW

File: ast/compile.go [hunk 1]
```diff
@@ -351,6 +356,12 @@ func (c *Compiler) Compile(modules map[string]*Module) {
 	c.compile()
 }
 
+// WithSchema sets a schema to the compiler
+func (c *Compiler) WithSchema(schema interface{}) *Compiler {
```
Comment (tsandall) [inline ast/compile.go:360]: We've talked about having OPA accept a set of schemas to help with use cases like Kubernetes where the source schemas are spread across many files ([example](https://github.com/garethr/kubernetes-json-schema)). We could generalize this to accept a set of schemas. Something like:

```golang
type SchemaSet struct {
   Schemas map[string][]byte
}

func (c *Compiler) WithSchemas(schema SchemaSet) *Compiler { ... }
```

This begs the question of how we relate schemas to `input` or other documents but we can defer that for now w/o worrying about having to change this interface.

File: ast/compile.go [hunk 2]
```diff
@@ -844,12 +855,102 @@ func (c *Compiler) checkSafetyRuleHeads() {
 	}
 }
 
+//CompileSchemas takes a jsonschema and compiles it.
+func CompileSchemas(byteSchema []byte, goSchema interface{}) (*gojsonschema.Schema, error) {
```
Comment (tsandall) [inline ast/compile.go:859]: I think we can hide this function as an implementation detail of the compilation process. Basically the compiler should accept a set of raw schemas (via `WithSchemas` above) and then when the type checking stage is executed, we can parse and compile the raw schemas, translate them into the type system, and add them into the environment.

File: ast/compile.go [hunk 3]
```diff
@@ -1348,6 +1449,7 @@ type queryCompiler struct {
 	after                map[string][]QueryCompilerStageDefinition
 	unsafeBuiltins       map[string]struct{}
 	comprehensionIndices map[*Term]*ComprehensionIndex
+	schema               interface{}
```
Comment (tsandall) [inline ast/compile.go:1452]: The `checkTypes` function on the query compiler should be extended to do the same thing as the Compiler.

File: bundle/store.go [hunk 4]
```diff
@@ -234,6 +234,7 @@ type ActivateOpts struct {
 	ExtraModules map[string]*ast.Module // Optional
 
 	legacy bool
+	Schema interface{}
```
Comment (tsandall) [inline bundle/store.go:237]: Since the compiler is provided as a parameter to the activate call, I think we may as well configure the schemas on it before calling activate. This file/package are complicated due to legacy reasons--I'd like to avoid changing it if possible.

File: bundle/store.go [hunk 5]
```diff
@@ -484,9 +485,14 @@ func writeModules(ctx context.Context, store storage.Store, txn storage.Transact
 		}
 	}
 
+	if schema != nil {
+		compiler = compiler.WithSchema(schema)
```
Comment (tsandall) [inline bundle/store.go:489]: Per comment above, we can do this outside.

File: cmd/eval.go [hunk 6]
```diff
@@ -195,6 +196,13 @@ Set the output format with the --format flag.
 	--format=values    : output line separated JSON arrays containing expression values
 	--format=bindings  : output line separated JSON objects containing variable bindings
 	--format=pretty    : output query results in a human-readable format
+
+Schema
+--------------
+
+The --schema flag will upload a single schema for the input document in JSON Schema format.
```
Comment (tsandall) [inline cmd/eval.go:203]: nit: I'd reword this a bit:

```
The -s/--schema flag provides a single JSON Schema used to validate references to the input document.
```

File: cmd/eval.go [hunk 7]
```diff
@@ -408,6 +417,17 @@ func setupEval(args []string, params evalCommandParams) (*evalContext, error) {
 		regoArgs = append(regoArgs, rego.ParsedInput(inputValue))
 	}
 
+	schemaBytes, err := readSchemaBytes(params)
+	if err != nil {
+		return nil, err
+	} else if schemaBytes != nil {
+		schema, err := ast.CompileSchemas(schemaBytes, nil)
```
Comment (tsandall) [inline cmd/eval.go:424]: Per comment in ast/compile.go, let's hide this as an implementation detail of the compiler's type checking stage. Just pass the raw schemas thru to the compiler.

**Review by tsandall — COMMENTED** [review]
This looks like a great start! I've left a few comments on the implementation. I think we should discuss whether/how OPA should accept multiple schemas at once--I think we have a few use cases where that'll be relevant--Kubernetes being the most obvious one.

File: topdown/query.go [hunk 8]
```diff
@@ -242,6 +243,13 @@ func (q *Query) WithResolver(ref ast.Ref, r resolver.Resolver) *Query {
 	return q
 }
 
+// WithSchema sets the schema object to use for the query. References rooted at
+// schema will be evaluated against this value. This is optional.
+func (q *Query) WithSchema(schema interface{}) *Query {
```
[the remark this answers is not in the document]
Comment (vazirim) [reply topdown/query.go]: Done (commit: https://github.com/open-policy-agent/opa/pull/3060/commits/9b7a0f960074b5247a0056e24f4c8e3a7427871d)

File: cmd/eval.go [hunk 9]
```diff
@@ -195,6 +196,13 @@ Set the output format with the --format flag.
 	--format=values    : output line separated JSON arrays containing expression values
 	--format=bindings  : output line separated JSON objects containing variable bindings
 	--format=pretty    : output query results in a human-readable format
+
+Schema
+--------------
```
[the remark this answers is not in the document]
Comment (vazirim) [reply cmd/eval.go]: Done (commit: https://github.com/open-policy-agent/opa/pull/3060/commits/9b7a0f960074b5247a0056e24f4c8e3a7427871d)

File: docs/content/schemas.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
+# Using schemas to enhance the Rego type checker
```
[the remark this answers is not in the document]
Comment (vazirim) [reply docs/content/schemas.md]: Done (commit: https://github.com/open-policy-agent/opa/pull/3060/commits/bebad7a80225206ce164a55efc1da8aa884615e6)

File: cmd/eval.go (hunk 7 above)
Comment (aavarghese) [reply cmd/eval.go]: Done

File: cmd/eval.go (hunk 6 above)
Comment (aavarghese) [reply cmd/eval.go]: Done

File: bundle/store.go (hunk 4 above)
Comment (aavarghese) [reply bundle/store.go]: Done

File: bundle/store.go (hunk 5 above)
Comment (aavarghese) [reply bundle/store.go]: Done

File: ast/compile.go (hunk 3 above)
Comment (aavarghese) [reply ast/compile.go]: Done

File: ast/compile.go (hunk 2 above)
Comment (aavarghese) [reply ast/compile.go]: Done

File: ast/compile.go (hunk 1 above)
Comment (vazirim) [reply ast/compile.go]: Done (commit https://github.com/open-policy-agent/opa/pull/3060/commits/174a097b64232c08ad7a7ee79d62bd5da9380789)

File: cmd/eval.go [hunk 11]
```diff
@@ -196,6 +197,13 @@ Set the output format with the --format flag.
 	--format=values    : output line separated JSON arrays containing expression values
 	--format=bindings  : output line separated JSON objects containing variable bindings
 	--format=pretty    : output query results in a human-readable format
+
+Schema
+------
+
+The -s/--schema flag provides a single JSON Schema used to validate references to the input document.
+
+	$ opa eval --data policy.rego --input input.json --schema input-schema.json
```
Comment (tsandall) [inline cmd/eval.go:206]: nit: I'm inclined to rename this to `--input-schema` so that we don't run into naming issues in the future when we want to support loading of multiple schemas.

File: docs/content/schemas.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
+---
+title: Type Checking
+kind: misc
+weight: 2
+---
+
+## Using schemas to enhance the Rego type checker
+
+You can provide an input schema to `opa eval` to improve static type checking and get more precise error reports as you develop Rego code.
+The `-s` flag can be used to upload a single schema for the input document in JSON Schema format.
```
Comment (tsandall) [inline docs/content/schemas.md:10]: Nit: since we'll want to rename this flag to avoid future naming conflicts, let's not include the short version of the flag.

File: rego/rego.go [hunk 13]
```diff
@@ -1044,6 +1052,10 @@ func New(options ...func(r *Rego)) *Rego {
 			WithBuiltins(r.builtinDecls)
 	}
 
+	if r.schemaSet != nil {
+		r.compiler.WithSchemas(r.schemaSet)
```
Comment (tsandall) [inline rego/rego.go:1056]: Move the `WithSchemas` call up to L1050 where the compiler is constructed. The reason is that if users pass their own compiler, we don't want to modify the schemas here. In other words, it's up to the caller that creates to the compiler to set the schemas to use for compilation.

**Review by tsandall — COMMENTED** [review]
Just two small comments and then I think we're pretty close. @srenatus can you take a look at the changes next week? Thanks!

File: cmd/eval.go (hunk 11 above)
Comment (vazirim) [reply cmd/eval.go:206]: @tsandall do you mean that in the future, we'll rename this flag (to upload schemas to be associated with something other than input)?

File: rego/rego.go (hunk 13 above)
Comment (vazirim) [reply rego/rego.go]: Done (commit https://github.com/open-policy-agent/opa/pull/3060/commits/31f947dbded99ceb8b83440a83323a544275d71a)

File: ast/compile.go [hunk 14]
```diff
@@ -844,12 +858,114 @@ func (c *Compiler) checkSafetyRuleHeads() {
...
+	if !ok {
+		return nil, fmt.Errorf("unexpected schema type %v", subSchema)
+	}
+
+	// Handle referenced schemas, returns directly when a $ref is found
+	if subSchema.RefSchema != nil {
+		return parseSchema(subSchema.RefSchema)
+	}
+
+	if subSchema.Types.IsTyped() && subSchema.Types.Contains(`boolean`) {
```
Comment (srenatus) [inline ast/compile.go:888]: 💭 Whenever I see strings with backticks, I wonder what's special about them. Here, nothing is, so I think we could also go with `"boolen"`, `"string"` etc. Anyways, it's just nitpicking. Don't bother 😄

Comment (srenatus) [inline ast/compile.go:888]: All these branches start with `subSchema.Types.IsTyped()`, what do you think about going with nested conditionals instead?
```go
if subSchema.Types.IsTyped() {
  if subSchema.Types.Contains("boolean") {
    return types.B, nil
  }
  if subSchema.Types.Contains("string") {
    return types.S, nil
  }
  // ...
}
```
not that we don't need the `else` if we have `return`s in the conditonals before.

File: ast/compile.go [hunk 15]
```diff
@@ -844,12 +858,114 @@ func (c *Compiler) checkSafetyRuleHeads() {
...
+	} else if subSchema.Types.IsTyped() && subSchema.Types.Contains(`integer`) {
+		return types.N, nil
+
+	} else if subSchema.Types.IsTyped() && subSchema.Types.Contains(`object`) {
+		if subSchema.PropertiesChildren != nil && len(subSchema.PropertiesChildren) > 0 {
+			staticProps := make([]*types.StaticProperty, 0, len(subSchema.PropertiesChildren))
+			for _, pSchema := range subSchema.PropertiesChildren {
+				newtype, err := parseSchema(pSchema)
+				if err != nil {
+					return nil, fmt.Errorf("unexpected schema type %v", pSchema)
```
Comment (srenatus) [inline ast/compile.go:903]: Is `err` interesting at all here? If so, let's do
```suggestion
					return nil, fmt.Errorf("unexpected schema type %v: %w", pSchema, err)
```
...but it might not be interesting? You'd know better 😃

File: cmd/eval.go [hunk 16]
```diff
@@ -410,6 +419,19 @@ func setupEval(args []string, params evalCommandParams) (*evalContext, error) {
 		regoArgs = append(regoArgs, rego.ParsedInput(inputValue))
 	}
 
+	schemaBytes, err := readSchemaBytes(params)
+	if err != nil {
+		return nil, err
+	} else if schemaBytes != nil {
```
Comment (srenatus) [inline cmd/eval.go:425]: Code-style-wise, I'd go with
```suggestion
	}
	if schemaBytes != nil {
```

File: go.sum [hunk 17]
```diff
@@ -296,6 +296,10 @@ github.com/urfave/cli v1.22.1/go.mod h1:Gos4lmkARVdJ6EkW0WaNv/tZAAMe9V7XWyB60NtX
 github.com/urfave/cli/v2 v2.1.1/go.mod h1:SE9GqnLQmjVa0iPEY0f1w3ygNIYcIJ0OKPMoW2caLfQ=
 github.com/wasmerio/go-ext-wasm v0.3.1 h1:G95XP3fE2FszQSwIU+fHPBYzD0Csmd2ef33snQXNA5Q=
 github.com/wasmerio/go-ext-wasm v0.3.1/go.mod h1:VGyarTzasuS7k5KhSIGpM3tciSZlkP31Mp9VJTHMMeI=
+github.com/xeipuuv/gojsonpointer v0.0.0-20190905194746-02993c407bfb h1:zGWFAtiMcyryUHoUjUJX0/lt1H2+i2Ka2n+D3DImSNo=
+github.com/xeipuuv/gojsonpointer v0.0.0-20190905194746-02993c407bfb/go.mod h1:N2zxlSyiKSe5eX1tZViRH5QA0qijqEDrYZiPEAiq3wU=
+github.com/xeipuuv/gojsonreference v0.0.0-20180127040603-bd5ef7bd5415 h1:EzJWgHovont7NscjpAxXsDA8S8BMYve8Y5+7cuRE7R0=
+github.com/xeipuuv/gojsonreference v0.0.0-20180127040603-bd5ef7bd5415/go.mod h1:GwrjFmJcFw6At/Gs6z4yjiIwzuJ1/+UwLxMQDVQXShQ=
```
Comment (srenatus) [inline go.sum:302]: I'm aware that I haven't been tracking the discussions that have happened about this before, so please bear with me -- I'm surprised about our vendoring choice here: We're manually vendoring the `github.com/xeipuuv/gojsonschema` code into `internal/`, and automatically vendor its dependencies via our ordinary means.

So... I'd think it would be more straightforward to only use the vendoring mechanisms we've been using for all other dependencies, i.e., not treat gojsonschema any special. This is assuming we don't need any changes to it -- maybe that's the reason and I'm missing the point here. _If that's the case_, let's drop a README into `internal/gojsonschem` to allow future readers to figure out why stuff was done the way it was done 😃

Comment (srenatus) [reply go.sum:302]: (Or rather `github.com/johandorland/gojsonschema`, it seems.)

**Review by srenatus — COMMENTED** [review]
Some code-style nitpicks and a general question re: vendoring 🙃 👇 please have a look.

File: ast/compile.go [hunk 18]
```diff
@@ -844,12 +858,114 @@ func (c *Compiler) checkSafetyRuleHeads() {
...
+	sl := gojsonschema.NewSchemaLoader()
+
+	if goSchema != nil {
+		refLoader = gojsonschema.NewGoLoader(goSchema)
+	} else {
+		return nil, fmt.Errorf("no schema as input to compile")
+	}
+	schemasCompiled, err := sl.Compile(refLoader)
+	if err != nil {
+		return nil, fmt.Errorf("unable to compile the schema due to: %s", err.Error())
```
[the remark this answers is not in the document]
Comment (vazirim) [reply ast/compile.go]: Done (commit https://github.com/open-policy-agent/opa/pull/3060/commits/2616431346afb485dd41e56d2b5f68100fb41d87)

File: ast/compile.go (hunk 14 above)
Comment (vazirim) [reply ast/compile.go]: Done (commit https://github.com/open-policy-agent/opa/pull/3060/commits/2616431346afb485dd41e56d2b5f68100fb41d87)

Comment (vazirim) [reply ast/compile.go]: Done (commit https://github.com/open-policy-agent/opa/pull/3060/commits/2616431346afb485dd41e56d2b5f68100fb41d87)

File: cmd/eval.go (hunk 11 above)
Comment (aavarghese) [reply cmd/eval.go:206]: Also, @tsandall based on our plan for the next upcoming PR (containing support for a. uploading multiple schemas via file(s) and directory path(s) and b. using rule based annotations with those in-memory schemas), our plan was to **re-use** this `opa eval --schema` argument. But the description for --schema in PR #2 would look something like this:

```--schema string                      set schema file(s) or directory path(s). This flag can be repeated.```

Given this next change ^^^, do you think we should leave --schema as-is, in this PR?

File: ast/compile.go (hunk 15 above)
Comment (vazirim) [reply ast/compile.go]: Done (commit https://github.com/open-policy-agent/opa/pull/3060/commits/2616431346afb485dd41e56d2b5f68100fb41d87)

File: cmd/eval.go (hunk 16 above)
Comment (vazirim) [reply cmd/eval.go]: Done (commit https://github.com/open-policy-agent/opa/pull/3060/commits/2616431346afb485dd41e56d2b5f68100fb41d87)

File: docs/content/schemas.md (hunk 12 above)
Comment (aavarghese) [reply docs/content/schemas.md:10]: Same as above comment ^^.

File: go.sum (hunk 17 above)
Comment (vazirim) [reply go.sum:302]: Yes we have made a modification to the gojsonschema library. We need to use the result of the parser to traverse it and obtain a Rego type. The library was not originally designed for this use case, and the data structure that represents a parsed schema has private fields. We made these fields public. There are other users of this library who have asked for this modification over the years, but they don't seem inclined to accept a PR.

The other modifications have to do with linting and conforming to the code style.

We will add a README to this effect.

Comment (aavarghese) [reply go.sum:302]: @srenatus thank you for the review and finding nits. Very helpful! Re: `gojsonschema` as you guessed, the reason we manually vendored into `internal/` was because of changes we had to do to expose some of the schema and sub-schema fields relevant for our type-checking code in OPA. Will add a README to explain this for future users.

Also, the

File: cmd/eval.go (hunk 11 above)
Comment (tsandall) [reply cmd/eval.go:206]: If we leave the flag as-is then what's the behaviour going to be in the next PR?

* if a single file path is passed, the schema applies to `input` in whole.
* if a directory path is passed, the schemas do not necessarily apply to `input` in whole; something else is needed to associate schemas with `input` or subpaths under `input` or `data`.

I suppose that's a reasonable default. In most cases, people will want to apply schema to `input` so the flag does that in the simplest way possible. However, it can be overloaded to support more sophisticated use cases as well.

OK, I'm fine with it as-is :)

File: go.sum (hunk 17 above)
Comment (aavarghese) [reply go.sum:302]: Done (commit 497cf35c844b333ec7e4fae74e375f22f035d083)

**Comment by aavarghese** [comment]
@tsandall @srenatus All comments have been addressed and changes committed. Thank you.

File: ast/compile_test.go [hunk 19]
```diff
@@ -4252,3 +4265,279 @@ func TestCompilerPassesTypeCheckNegative(t *testing.T) {
...
+		t.Fatalf("unexpected error: %s", err)
+	}
+	newtype, err := setTypesWithSchema(sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	if newtype == nil {
+		t.Fatalf("parseSchema returned nil type")
+	}
+	if newtype.String() == "object<apiVersion: string, kind: string, metadata: any, spec: any, status: any>" {
```
Comment (srenatus) [inline ast/compile_test.go:4304]: Can we assert the expected type instead?

Relatedly, is there a way to have a structured comparison instead of comparing strings?

File: ast/compile_test.go [hunk 20]
```diff
@@ -4252,3 +4265,279 @@ func TestCompilerPassesTypeCheckNegative(t *testing.T) {
...
+	}
+	newtype, err := parseSchema(jsonSchema) // Did not pass the subschema
+	if newtype != nil {
+		t.Fatalf("Incorrect return from parseSchema with a bad schema")
+	}
+}
+
+func TestWithSchema(t *testing.T) {
+	c := NewCompiler().
+		WithCapabilities(&Capabilities{Builtins: []*Builtin{Split}})
```
Comment (srenatus) [inline ast/compile_test.go:4357]: I don't yet understand the significance of having this builtin here...?

**Review by srenatus — COMMENTED** [review]
Thanks for addressing my concerns, that README will sure help in the future.

I have yet to play with this locally, and added a few more comments in the tests of the compiler.

Overall, it's a great feature. Looking forward to see this widely used 😄

File: ast/compile_test.go (hunk 20 above)
Comment (vazirim) [reply ast/compile_test.go:4357]: Not needed indeed. Will remove, thanks.

Comment (vazirim) [reply ast/compile_test.go]: Done  (commit https://github.com/open-policy-agent/opa/pull/3060/commits/bd984a0fcc39b763449b8ea3616d419eeb925ab5)

File: ast/compile_test.go (hunk 19 above)
Comment (vazirim) [reply ast/compile_test.go:4334]: Fixed in commit https://github.com/open-policy-agent/opa/pull/3060/commits/bd984a0fcc39b763449b8ea3616d419eeb925ab5

@srenatus  I changed the comparisons to use `Compare` on types instead of string comparisons. 

The test above is for a schema that contains many references. So when all of those are integrated and parsed, the resulting Rego type is ~100 lines long (when printed on my terminal), so it's not very convenient to have a precise assertion. This test is asserting that we don't get any errors when we parse such a schema, and that the newtype is not the one we would have obtained if we had ignored the refs.

I included another tests for Schema refs that is much shorter, and in that case we have a string comparison for convenience to the expected type.

Pl let me know what you think.

Comment (srenatus) [reply ast/compile_test.go:4334]: Thanks, this is fine 👍 

The crucial, to me, bits are these:
> the newtype is not the one we would have obtained if we had ignored the refs.

from reading the code, I didn't understand the meaning of that string it wasn't supposed to be.

File: docs/content/schemas.md [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+```
+
+Example:
+```
+opa eval data.envoy.authz.allow -i example/envoy/input.json -d example/envoy/policy.rego -s example/envoy/input-schema.json
+```
+Samples provided at: https://github.com/aavarghese/opa-schema-examples/tree/main/envoy
+
+
+{{< figure src="OPATypeChecking_PR3053.gif" width="130" caption="Demo" >}}
```
Comment (srenatus) [inline docs/content/schemas.md:23]: This will become part of the docs page, I think. Can we drop the gif animation, please? I think the _Usage Scenario_ outlined below is a nice, self-contained introduction. The page could then end with a _reference_ section pointing towards
- the opa-schema-examples repo,
- and maybe some official jsonschema docs page?

Comment (srenatus) [reply docs/content/schemas.md:23]: I do think that should be publicised widely, and the gif example could help for quickly grasping what this is about. But I think I'd prefer it not in the docs... What do you think? 😃

**Review by srenatus — COMMENTED** [review]
ℹ️ Played with this locally, it's pretty great 😄 

One last nitpick in the docs, then I'd be happy to merge this, and let us iterate from here.

File: docs/content/schemas.md (hunk 21 above)
Comment (vazirim) [reply docs/content/schemas.md]: Done (commit https://github.com/open-policy-agent/opa/pull/3060/commits/0e0d9152740e6f6d9035a6bb6250ee0bf24d493e)

Perhaps the gif could be useful for release notes or we could include it in a blog...

Comment (srenatus) [reply docs/content/schemas.md]: I think we should definitely do a blog post here! 🚀

**Comment by tsandall** [comment]
@aavarghese I don't see anything else that needs to be changed. Can you squash your commits? One commit w/ a comprehensive message explaining the changes (and any subtleties like the forked/copied JSON schema library) would be fine.

**Comment by aavarghese** [comment]
> @aavarghese I don't see anything else that needs to be changed. Can you squash your commits? One commit w/ a comprehensive message explaining the changes (and any subtleties like the forked/copied JSON schema library) would be fine.

@tsandall Commits have been squashed with a message. Thanks.

**Review by srenatus — APPROVED** [review]
🚀 🎉 👏

**Comment by srenatus** [comment]
ℹ️ I've taken the liberty to copy-edit the commit message a little bit, hope you don't mind. Your content is intact, I've added a crisp subject line (the first line, followed by a blank one) and a few line breaks.

**Comment by aavarghese** [comment]
> ℹ️ I've taken the liberty to copy-edit the commit message a little bit, hope you don't mind. Your content is intact, I've added a crisp subject line (the first line, followed by a blank one) and a few line breaks.

Thanks @srenatus. All looks good. Appreciate all the thorough reviewing and detailed feedback. :)
